cloudhbase-connector/src/main/java/com/alibaba/flink/connectors/cloudhbase/sink/CloudHBaseSinkFunction.java

```diff
@@ -0,0 +1,342 @@
+	}
+
+	public CloudHBaseSinkFunction(
+			String zkQuorum,
+			String zkZNodeParent,
+			String tableName,
+			int bufferSize,
+			int batchSize,
+			long flushInterval,
+			int maxRetryNumber,
+			CloudHBaseRecordResolver<RECORD> recordResolver) {
+		this.zkQuorum = zkQuorum;
+		this.zkZNodeParent = zkZNodeParent;
+		this.tableName = tableName;
+		this.bufferSize = bufferSize;
+		this.batchSize = batchSize;
+		this.flushInterval = flushInterval;
+		this.maxRetryNumber = maxRetryNumber;
+		this.recordResolver = recordResolver;
+	}
+
+	public String getZkQuorum() {
+		return zkQuorum;
+	}
+
+	public String getZkZNodeParent() {
+		return zkZNodeParent;
+	}
+
+	public String getTableName() {
+		return tableName;
+	}
+
+	public int getMaxRetryNumber() {
+		return maxRetryNumber;
+	}
+
+	public int getBufferSize() {
+		return bufferSize;
+	}
+
+	public int getBatchSize() {
+		return batchSize;
+	}
+
+	public long getFlushInterval() {
+		return flushInterval;
+	}
+
+	public CloudHBaseRecordResolver getCloudHBaserecordResolver() {
+		return recordResolver;
+	}
+
+	@Override
+	public void open(Configuration config) throws IOException {
+		this.recordResolver.open();
+
+		// create hbase connection
+		org.apache.hadoop.conf.Configuration conf = HBaseConfiguration.create();
+		conf.set(HConstants.ZOOKEEPER_QUORUM, zkQuorum);
+		if (!StringUtils.isNullOrWhitespaceOnly(zkZNodeParent)) {
+			conf.set(HConstants.ZOOKEEPER_ZNODE_PARENT, zkZNodeParent);
+		}
+		conf.setInt(HConstants.ZK_SESSION_TIMEOUT, 30000);
+		conf.setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, maxRetryNumber);
+		conf.set(HConstants.RPC_CODEC_CONF_KEY, "");
+		this.hbaseConn = ConnectionFactory.createConnection(conf);
+
+		this.writeBuffer = new HashMap<>();
+		this.lastFlushTime = 0;
+		this.flushError = false;
+
+		startFlusher();
+
+		this.outTps = MetricUtils.registerOutTps(getRuntimeContext());
+		this.outBps = MetricUtils.registerOutBps(getRuntimeContext(), "cloudhbase");
+		this.latencyGauge = MetricUtils.registerOutLatency(getRuntimeContext());
+
+		LOG.info("CloudHBase is opened");
+	}
+
+	private void startFlusher() {
+		this.flusher = new ScheduledThreadPoolExecutor(1,
+				new DispatcherThreadFactory(Thread.currentThread().getThreadGroup(), "CloudHBase Flusher"));
+		this.flusher.setRemoveOnCancelPolicy(true);
+		this.flusher.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
+		this.flusher.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+
+		this.flusher.scheduleAtFixedRate(() -> {
+			if (System.currentTimeMillis() - lastFlushTime >= flushInterval) {
+				flush(false);
+			}
+		}, flushInterval, flushInterval, TimeUnit.MILLISECONDS);
+	}
+
+	@Override
+	public void invoke(RECORD record) throws IOException {
+		if (flushError) {
+			throw new RuntimeException(flushException);
+		}
+
+		String rowKey = recordResolver.getRowKey(record);
+		synchronized (writeBuffer) {
+			writeBuffer.put(rowKey, record);
+		}
+
+		if (writeBuffer.size() >= bufferSize) {
```

Comment:
Should the size check also be in synchronized block?

Reply:
The only race condition here is that we get `writeBuffer.size()` just before a periodically `flush()`, and the `if` is `True`，so a `flush()` will be triggered here after the periodically `flush()` which actually should not happen. But in `flush()`, the `writeBuffer.size()` will be rechecked again, and just return in this case. I think this race condition is rare, and there is no need to put in synchronized block which can do a more fine-grained synchronization. On the other hand, `invoke` is called much more frequently than `flush()`, so I think it's better to recheck the `writeBuffer.size()` in the synchronized block of `flush` instead of `invoke`.